src/storage/src/upsert_continual_feedback.rs

```diff
@@ -0,0 +1,826 @@
+                                // input.
+                                //
+                                // This is a load-bearing optimization, as it is
+                                // required to avoid buffering the entire source
+                                // snapshot in the `stash`.
+                                if prevent_snapshot_buffering && resume_upper.less_equal(&G::Timestamp::minimum()) && event_time == G::Timestamp::minimum() {
+                                    tracing::info!(?event_time, ?resume_upper, ?output_cap, "partial drain not implemented, yet");
+                                }
+                            }
+                            AsyncEvent::Progress(upper) => {
+                                tracing::trace!(?upper, "received progress in upsert");
+
+                                // Ignore progress updates before the `resume_upper`, which is our initial
+                                // capability post-snapshotting.
+                                if PartialOrder::less_than(&upper, &resume_upper) {
+                                    tracing::trace!(?upper, ?resume_upper, "ignoring progress updates before resume_upper");
+                                    continue;
+                                }
+
+                                if let Some(ts) = upper.as_option() {
+                                    tracing::trace!(?ts, "downgrading output capability");
+                                    let _ = output_cap.try_downgrade(ts);
+                                }
+                                input_upper = upper;
+                            }
+                        }
+                        let events_processed = i + 1;
+                        if let Some(max) = snapshot_buffering_max {
+                            if events_processed >= max {
+                                break;
+                            }
+                        }
+                    }
+                }
+            };
+
+            // We try and drain from our stash every time we go through the
+            // loop. More of our stash can become eligible for draining both
+            // when the source-input frontier advances or when the persist
+            // frontier advances.
+
+            drain_staged_input::<_, G, _, _, _>(
+                &mut stash,
+                &mut commands_state,
+                &mut output_updates,
+                &mut multi_get_scratch,
+                DrainStyle::ToUpper{input_upper: &input_upper, persist_upper: &persist_upper},
+                &mut error_emitter,
+                &mut state,
+            )
+            .await;
+
+            tracing::trace!(?output_updates, "output updates for complete timestamp");
+
+            // TODO: Feels somewhat inefficient to be peeling
+            // of and creating new vecs per group of
+            // timestamp. Especially since at steady state
+            // we're expecting to only have one group...
```

Comment:
This is a totally ordered timestamp operator so we should hold onto a single capability and drive it around. Emitting the data below can be done with a single capability without necessarily having the DD time of the triplet match the time of the capability.

Also, it looks like we are creating a clone of the capability per triplet. This can lead to a lot of time being spend in ChangeBatch maintenance. This is another reason to have `output_updates` only contain the triplets to be emitted

Reply:
I chose to do it like this because its the most conservative option: because whether we can emit updates does not depend only on the input frontier, but also on the persist input frontier, we can't just downgrade our output_cap when we get progress updates on the regular input. I didn't want to futz with that, and saw some hard-to-debug things while I was developing.

Just keeping the input cap means we can always emit with that. We might want to look into optimizing that later on, though.